### ISSUE TITLE
[BUG] PixbufAnimation.get_static_image should not return Option<Pixbuf>

### ISSUE THREAD

**Issue by aruiz** [gtk-rs/gtk-rs-core#922]
As per the docs get_static_image() returns a (transfer none) Pixbuf or none if the data cannot be loaded yet, meaning it is owned by the PixbufAnimation.

My take is that we should return Option<&Pixbuf> as, besides being nullable, the returned Pixbuf is tied to the PixbufAnimation's lifetime.

**Comment by sdroege** [comment]
It already returns an `Option`.

For returning a borrowed reference (`&`) instead of a new owned reference, we can only do that if the value can never change for the same instance. Is that the case here, i.e. will it always return `None` up to a certain point and from then on always return exactly the same `Pixbuf`?

Also, please specify if you mean the function in the `PixbufAnimationExt` or `PixbufAnimationImpl` trait :)

**Comment by aruiz** [comment]
> It already returns an `Option`.
> 
> For returning a borrowed reference (`&`) instead of a new owned reference, we can only do that if the value can never change for the same instance. Is that the case here, i.e. will it always return `None` up to a certain point and from then on always return exactly the same `Pixbuf`?

Yes, this is basically a preview frame (or full size thumbnail) generated from the image data, the Animation may not have enough data yet to build that image, or either the data is corrupt and it can't generate the thumbnail. But if the data is valid it will generate the thumbnail just once.

PixbufAnimation is not designed so that the user can update the image data once it was created, so no new thumbnail will be needed once the first one is created as a Pixbuf instance. (An implementation may choose to not respect this intent by adding a function that allows updating the data do this but it will certainly be an ugly hack that goes outside of that the API is designed for).

So tl;dr, yes, it will only generate either none or one instance for the lifetime of the PixbufAnimation.

> Also, please specify if you mean the function in the `PixbufAnimationExt` or `PixbufAnimationImpl` trait :)

Sorry but I am still getting a handle as to what each one is for :D

**Comment by sdroege** [comment]
> So tl;dr, yes, it will only generate either none or one instance for the lifetime of the PixbufAnimation.

Can you confirm that none of the normal pixbuf loaders is returning a different pixbuf when the function is called and it's indeed tied to the lifetime of the object?

And also that if a pixbuf loader actually does that then it's considered a bug?

If so, we can change this. Might also make sense to change the two `gio::IOStream` getters the same way.

> Sorry but I am still getting a handle as to what each one is for :D

The former is what the user/application calls, the latter is what you implement to implement a new `PixbufAnimation` :)

Now for the latter, if it was returning an `Option<&Pixbuf>`, I assume you would have problems because you probably store it in a `RefCell<&Pixbuf>` and returning a `&Pixbuf` from that is going to need `unsafe` code unless I need more coffee first. That would seem suboptimal.

Using `OnceCell<Pixbuf>` would seem like a safe option for that though and I think that would actually be a good idea.

Any thoughts?

**Comment by aruiz** [comment]
> Can you confirm that none of the normal pixbuf loaders is returning a different pixbuf when the function is called and it's indeed tied to the lifetime of the object?

Ugh... it's all over the place actually. Some implementations behave according to the documentation and some don't. This is memory leak/double free galore...

> And also that if a pixbuf loader actually does that then it's considered a bug?

Lemme file a bug against those (and possibly a PR) and see what the maintainer says.
 
> If so, we can change this. Might also make sense to change the two gio::IOStream getters the same way.
The former is what the user/application calls, the latter is what you implement to implement a new PixbufAnimation :)

Let's see what input we get from the maintainers, I will post any updates.

> Now for the latter, if it was returning an Option<&Pixbuf>, I assume you would have problems because you probably store it in a RefCell<&Pixbuf> and returning a &Pixbuf from that is going to need unsafe code unless I need more coffee first. That would seem suboptimal.
> 
> Using OnceCell<Pixbuf> would seem like a safe option for that though and I think that would actually be a good idea.
> 
> Any thoughts?

How you would distribute several references to the same Pixbuf using OnceCell without copying the Pixbuf? I was thinking more something along the lines of Rc<Pixbuf>

**Comment by sdroege** [comment]
Sounds fun! Let's see what the results of that discussion is. Can you provide a link to the issue(s)?

> How you would distribute several references to the same Pixbuf using OnceCell without copying the Pixbuf? I was thinking more something along the lines of `Rc<Pixbuf>`

Just clone the pixbuf then. It's a GObject so it's refcounted :)

**Comment by aruiz** [comment]
For the record:
https://gitlab.gnome.org/GNOME/gdk-pixbuf/-/issues/222

Turns out only the gif implementation violates the introspection annotation. Let's see what they say.

**Comment by sdroege** [comment]
As discussed on Matrix this should be fixed for the `Impl` trait as currently it's plain wrong and easily allows a use-after-free. For the `Ext` trait it doesn't really matter but would make sense for consistency.

**Comment by aruiz** [comment]
I just reviewed the whole Animation/Iter API. Iter.pixbuf() and Animation.static_image() are the only two callouts here.

Would it make sense to return Option<&Pixbuf> and &Pixbuf respectively? I can give it a go at writing the patch myself.

**Comment by sdroege** [comment]
I think so, yes. This needs to be changed in 6 places or so per function.

**Comment by aruiz** [comment]
I will give this a go in two separate PRs

**Comment by aruiz** [comment]
I just noticed PixbufLoader.pixbuf() has the same issue... will try to fix that one as well.

**Comment by aruiz** [comment]
Okay, after a conversation with @GuillaumeGomez I think it's best not to return a & as the lifetime is problematic, will just return a .clone() and live with the upref, my guess is that most bindings do this anyway.

**Comment by sdroege** [comment]
Even if we don't return a reference there is still a bug here in that you can return a pixbuf without keeping it alive.

**Comment by sdroege** [comment]
See https://github.com/gtk-rs/gtk-rs-core/pull/934 . I didn't return references after all because even then nothing would ensure that the subclass is keeping them alive long enough. So using the normal quark workaround here too.
